src/index.js

```diff
@@ -12,55 +12,145 @@
-        const ruleError = new RuleError(message, {
-          index: originalIndex,
-          fix
-        });
-        report(node, ruleError);
+const constructNSpell = function (language) {
+  const base = require.resolve(`dictionary-${language}`);
+  const dictionary = fs.readFileSync(
+    path.join(base, "..", "index.dic"),
+    "utf-8"
+  );
+  const affix = fs.readFileSync(path.join(base, "..", "index.aff"), "utf-8");
+  return nspell(affix, dictionary);
+};
+
+const getRuleErrorBuilder = (nSpell, suggestCorrections, RuleError, fixer) => {
+  if (suggestCorrections) {
+    return (word, originalRange) => {
+      const suggestions = nSpell.suggest(word);
+      const fix =
+        suggestions.length === 1
+          ? fixer.replaceTextRange(originalRange, suggestions[0])
+          : undefined;
+      const message = `${word} -> ${suggestions.join(", ")}`;
+      return new RuleError(message, {
+        index: originalRange[0],
+        fix: fix,
+      });
+    };
+  } else {
+    return (word, originalRange) => {
+      return new RuleError(word, {
+        index: originalRange[0],
+      });
+    };
+  }
+};
+
+const doesExclusionOverlapWord = (
+  [exclusionRangeStart, exclusionRangeEnd],
+  [wordRangeStart, wordRangeEnd]
+) => {
+  const startsBeforeEndOfExclusion = wordRangeStart <= exclusionRangeEnd;
+  const endsAfterStartOfExclusion = wordRangeEnd >= exclusionRangeStart;
+  return startsBeforeEndOfExclusion && endsAfterStartOfExclusion;
+};
+
+const getNoExclusionsOverlapWord = (exclusions, indexOffset) => {
+  if (exclusions && exclusions.length > 0) {
+    return (originalRange) => {
+      return !exclusions.some(({ startIndex, endIndex }) =>
+        doesExclusionOverlapWord(
+          [startIndex + indexOffset, endIndex + indexOffset],
+          originalRange
+        )
+      );
+    };
+  }
+  return (originalRange) => true;
+};
+
+const reporter = function (
+  { Syntax, RuleError, fixer, report },
+  {
+    language = "en",
+    skipPatterns = [],
+    wordDefinitionRegex: optionWordDefinitionRegex = /[\w']+/g,
+    suggestCorrections = true,
+  }
+) {
+  const nSpell = constructNSpell(language);
+
+  const ruleErrorBuilder = getRuleErrorBuilder(
+    nSpell,
+    suggestCorrections,
+    RuleError,
+    fixer
+  );
+
+  const handleText = (node, text, indexOffset) => {
+    if (!text) return;
+    const wordDefinitionRegex = new RegExp(optionWordDefinitionRegex);
+
+    let noExclusionsOverlapWord;
+
+    let matches;
+    while ((matches = wordDefinitionRegex.exec(text)) !== null) {
+      // This is necessary to avoid infinite loops with zero-width matches
+      if (matches.index === wordDefinitionRegex.lastIndex) {
+        wordDefinitionRegex.lastIndex++;
+      }
+      const originalIndex = indexOffset + matches.index;
+      const word = matches[0];
+
+      const originalWordRange = [originalIndex, originalIndex + word.length];
+
+      if (!nSpell.correct(word)) {
+        noExclusionsOverlapWord =
```

Comment:
You can move this code to line 102 with following declaration, and remove `let`.
```js
const noExclusionsOverlapWord = getNoExclusionsOverlapWord(matchPatterns(text, skipPatterns), indexOffset);
```
`text` and `skipPatterns`, `indexOffset` is defined outside of `while`-loop. Thus, you don't need to assign `noExclusionsOverlapWord` in this loop.

Reply:
Whilst true, the reason the this is split over 2 locations the 
```javascript
let noExclusionsOverlapWord;
```
ensures that we cache this function (which is specific to the current text part we are validating) once for this specific text part.

I put the `noExclusionsOverlapWord = ...` after `if (!nSpell.correct(word)) {` so we only have to calculate the possible exclusions for text parts containing an incorrect word. Which are the only text parts for whom the exclusions have any significance. So text parts without a spelling error are not "punished" by having to figure out possible exclusion ranges which will never be used, while text parts with multiple wrong words will still only calculate the exclusion ranges once. This is the optimization I talked about in my initial PR message.

So while it is true that it could be declared in the outer scope, it is more performant to do this conditionally in the inner scope.